astropy/units/tests/test_quantity_annotations.py

```diff
@@ -1,9 +1,112 @@
 # -*- coding: utf-8 -*-
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
 
+# STDLIB
+import typing as T
+
+# THIRD PARTY
 import pytest
 
+# LOCAL
 from astropy import units as u  # pylint: disable=W0611
+from astropy.units import Quantity
+from astropy.units._typing import HAS_ANNOTATED, Annotated
+
+
+def test_ignore_generic_type_annotations():
+    """Test annotations that are not unit related are ignored.
+
+    This test passes if the function works.
+    """
+    # one unit, one not (should be ignored)
+    @u.quantity_input
+    def func(x: u.m, y: str):
+        return x, y
+
+    i_q, i_str = 2 * u.m, "cool string"
+    o_q, o_str = func(i_q, i_str)  # if this doesn't fail, it worked.
+    assert i_q == o_q
+    assert i_str == o_str
+
+
+@pytest.mark.skipif(not HAS_ANNOTATED, reason="need `Annotated`")
+class TestQuantityUnitAnnotations:
+    """Test Quantity[Unit] type annotation."""
+
+    def test_simple_annotation(self):
+        @u.quantity_input
+        def func(x: Quantity[u.m], y: str):
+            return x, y
+
+        i_q, i_str = 2 * u.m, "cool string"
+        o_q, o_str = func(i_q, i_str)
+        assert i_q == o_q
+        assert i_str == o_str
+
+    def test_multiple_annotation(self):
+        @u.quantity_input
+        def multi_func(a: Quantity[u.km]) -> Quantity[u.m]:
+            return a.to(u.m)
+
+        i_q = 2 * u.km
+        o_q = multi_func(i_q)
+        assert o_q == i_q
+        assert o_q.unit == u.m
+
+    @pytest.mark.skipif(not HAS_ANNOTATED, reason="need `Annotated`")
+    def test_optional_and_annotated(self):
+        @u.quantity_input
+        def opt_func(x: T.Optional[Quantity[u.m]] = None) -> Quantity[u.km]:
+            if x is None:
+                return 1 * u.km
+            return x
+
+        i_q = 2 * u.m
+        o_q = opt_func(i_q)
+        assert o_q.unit.physical_type == "length"
+        assert o_q == i_q
```

Comment:
Might be good to check the output unit here? Also, since rounding can be an issue, maybe chose the input to something like `250*u.m`?

Reply:
SGTM.

Reply:
Done.